### PR TITLE
feat: add valuables plugin

### LINKED ISSUE #2599: [Bounty] Display v.cent current bid price of the tweet directly on Twitter
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [x] Beginner
  - [ ] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 1 week
- Bounty Size
  - [x] 500 USD + 500 USD matched by Cent Team!!!

## Part 1: Problem Description
Do I own my tweet? Actually we don't care about it. No matter if Twitter the platform "permits" it or not, Valuables by cent has created a way to help you "own" your Tweet and be able to sell it. Jack Dorsey(@jack@Twitter, founder of Twitter) also sold his first tweet on it (https://v.cent.co/tweet/20, https://twitter.com/jack/status/20). The first step of our cooperation with Valuables is to directly display the current bid price of each Tweet, if it has been offered a bid.
Use their API https://v.cent.co/data/tweet-txn?tweetID=<TWEETID> (e.g. https://v.cent.co/data/tweet-txn?tweetID=20) to get the current bid price and display it within each Tweet card. The UI is all up to you.

## Part 2: Resources
https://v.cent.co/
https://v.cent.co/data/tweet-txn?tweetID=<TWEETID>
https://docs.google.com/document/d/1iXhertpMT5Y8x-2HmemMtpNB6x9sWPUWJq8s-oacweQ/edit

## Part 3: Contact and Support
@yisiliu for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q

### PR DESCRIPTION
**Pull request by MountVase**
### Plugin to display potential tweet bids directly in Twitter post. 
fix #2599  

#### Screenshots
vcent_plugin_dark
vcent_plugin_light

### CODE REVIEW

File: packages/maskbook/src/plugins/Valuables/README.md
Comment (septs) [inline packages/maskbook/src/plugins/Valuables/README.md:3]:
```suggestion
Valuables by CENT <https://v.cent.co>
```

File: packages/maskbook/src/plugins/Valuables/README.md
Comment (septs) [inline packages/maskbook/src/plugins/Valuables/README.md:11]:
```suggestion
API endpoint: <https://v.cent.co/data/tweet-txn/latest?tweetID=[TWEETID]>
```

File: packages/maskbook/src/plugins/Valuables/README.md
Comment (septs) [inline packages/maskbook/src/plugins/Valuables/README.md:22]:
```suggestion
CORS problems with API. No Access-Control-Allow-Origin header returned.
Might be because I'm fetching from localhost, and not in production and from twitter.com.
Local development workaround: [Moesif CORS](https://chrome.google.com/webstore/detail/moesif-origin-cors-change/digfbfaphojjndkpccljibejjbppifbc)
```

**Review by septs — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/Valuables/api.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Valuables/api.tsx:22]: Maybe we need to add a proxy server for this. cc @yisiliu or @septs?

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:38]: Can be more clean:

```ts
() => getTweetBid(props.id)
```

File: packages/maskbook/src/plugins/Valuables/define.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Valuables/define.tsx:21]: No need to add successDecryptionInspector for this plugin

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:17]: hmm are you sure a button for 100 view port width?

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:46]:
```tsx
if (tweetBidResponse.loading || tweetBidResponse.error || !tweetBidResponse.value) return null
return <div> ... </div>
```

**Comment by Jack-Works** [comment]
Cool! I only have some style reviews. This PR LGTM. But I have a question, is it too expensive to check every tweet?

File: packages/maskbook/src/plugins/Valuables/api.tsx
Comment (MountVase) [reply packages/maskbook/src/plugins/Valuables/api.tsx:22]: Yeah I thought about that, looked at this one for a template: https://github.com/Rob--W/cors-anywhere.

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (MountVase) [reply packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx]: Changed it to 100% width, do you mean you're unsure of using the button component there or that it's too wide?

**Comment by MountVase** [comment]
> Cool! I only have some style reviews. This PR LGTM. But I have a question, is it too expensive to check every tweet?

Thanks! I have the same question.. Currently doesn't look that way to me but it surely takes up more resources than it optimally would. 
Thinking to maybe use useMemo() or some caching to at least make sure that we're not fetching for the same tweet several times.

**Comment by Jack-Works** [comment]
useMemo() doesn't work across different components. you'll need a global cache. and there's prettier error plz fix it thanks!

**Comment by MountVase** [comment]
Added Flags configuration. Let me know if it should be different than 'betaOrInsiderOnly'.

**Comment by Tedko** [comment]
cc @jackworks

**Review by Jack-Works — COMMENTED** [review]
Looks good to me! But before we merge this, we need to figure out how to do with the API CORS limit

File: packages/maskbook/src/plugins/Valuables/constants.ts
Comment (septs) [inline packages/maskbook/src/plugins/Valuables/constants.ts:1]:
```suggestion
// proxy: https://v.cent.co/data/tweet-txn
export const VCENT_BASE_URL = 'https://vcent-agent.r2d2.to/data/tweet-txn?tweetID='
```

File: packages/maskbook/src/plugins/Valuables/README.md
Comment (septs) [inline packages/maskbook/src/plugins/Valuables/README.md:17]: remove the section.

File: packages/maskbook/src/plugins/Valuables/api.tsx
Comment (septs) [inline packages/maskbook/src/plugins/Valuables/api.tsx:18]:
```suggestion
    return results[0]
```

Comment (septs) [reply packages/maskbook/src/plugins/Valuables/api.tsx:18]: rename the file name to `api.ts`

File: packages/maskbook/src/plugins/Valuables/api.tsx
Comment (septs) [reply packages/maskbook/src/plugins/Valuables/api.tsx]: plz use proxy service, replace it.

```
https://vcent-agent.r2d2.to/data/tweet-txn?tweetID=
```

**Review by Jack-Works — APPROVED** [review]

**Comment by septs** [comment]
plz resolve conflicts.

**Review by Jack-Works — DISMISSED** [review]

**Comment by Jack-Works** [comment]
merge conflicts resolved. Did anyone test with the functionality?

**Comment by yisiliu** [comment]
@MountVase Also some last minute review: can you please follow the design in the [figma link](https://www.figma.com/file/gVkQ67y285b4FXVV1KPThN/Twitter?node-id=8915%3A347) and adjust the color and the font size please?

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (MountVase) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:26]:
```suggestion
        color: '#1C68F3',
        fontSize: '14px',
```

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:6]: Remove unused code.

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:15]: `borderRadius: 20,`

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:27]: `fontSize: 14,`

File: packages/maskbook/src/plugins/Valuables/api.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Valuables/api.ts:5]: Does it strict enough? Or you maybe need `/^\d+$/`.

File: packages/maskbook/src/plugins/Valuables/api.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Valuables/api.ts:14]: Merge those lines to `first(results)`.

File: packages/maskbook/src/plugins/Valuables/api.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Valuables/api.ts:31]: The comment can be removed by using `latest: 0 | 1`.

File: packages/maskbook/src/plugins/Valuables/constants.ts
Comment (septs) [reply packages/maskbook/src/plugins/Valuables/constants.ts:2]: plz merge it.

**Review by Jack-Works — DISMISSED** [review]

**Comment by Tedko** [comment]
@Jack-Works @septs @MountVase guys whats the update so far? we wanna deliver this feature asap.

**Comment by Jack-Works** [comment]
Need one more reviewer. I have already approved. (Merge master to fix all CI errors).

File: packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx
Comment (UncleBill) [inline packages/maskbook/src/plugins/Valuables/UI/BidCard.tsx:49]: nit: `weetBidResponse.value` got nullable-checked above.
```suggestion
                    Bid: ${tweetBidResponse.value.amount_usd} ({tweetBidResponse.value.amount_eth} ETH)
```

File: packages/maskbook/src/plugins/Valuables/api.ts
Comment (UncleBill) [inline packages/maskbook/src/plugins/Valuables/api.ts:31]: nit: put `amount*`s closer together.

```suggestion
    amount_eth: number // eth
    amount_usd: number // usd
    txid: string
    chain_time: string
    latest: 0 | 1
    create_time: string
    status: string
```

**Review by Jack-Works — APPROVED** [review]

**Pull request closed without merging**
